File: lib/ruby/signature/ast/declarations.rb
module Ruby
  module Signature
    module AST
      module Declarations
        class Class
          class Super
            attr_reader :name
            attr_reader :args

            def initialize(name:, args:)
              @name = name
              @args = args
            end

            def ==(other)
              other.is_a?(Super) && other.name == name && other.args == args
            end

            alias eql? ==

            def hash
              self.class.hash ^ name.hash ^ args.hash
            end

            def to_json(*a)
              {
                name: name,
                args: args
              }.to_json(*a)
            end
          end

          attr_reader :name
          attr_reader :type_params
          attr_reader :members
          attr_reader :super_class
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, type_params:, super_class:, members:, annotations:, location:)
            @name = name
            @type_params = type_params
            @super_class = super_class
            @members = members
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(Class) &&
              other.name == name &&
              other.type_params == type_params &&
              other.super_class == super_class &&
              other.members == members
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type_params.hash ^ super_class.hash ^ member.hash
          end

          def to_json(*a)
            {
              declaration: :class,
              name: name,
              type_params: type_params,
              members: members,
              super_class: super_class,
              annotations: annotations,
              location: location
            }.to_json(*a)
          end
        end

        class Module
          attr_reader :name
          attr_reader :type_params
          attr_reader :members
          attr_reader :location
          attr_reader :annotations
          attr_reader :self_type

          def initialize(name:, type_params:, members:, self_type:, annotations:, location:)
            @name = name
            @type_params = type_params
            @self_type = self_type
            @members = members
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(Module) &&
              other.name == name &&
              other.type_params == type_params &&
              other.self_type == self_type &&
              other.members == members
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type_params.hash ^ self_type.hash ^ members.hash
          end

          def to_json(*a)
            {
              declaration: :module,
              name: name,
              type_params: type_params,
              members: members,
              self_type: self_type,
              annotations: annotations,
              location: location
            }.to_json(*a)
          end
        end

        class Extension
          attr_reader :name
          attr_reader :type_params
          attr_reader :extension_name
          attr_reader :members
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, type_params:, extension_name:, members:, annotations:, location:)
            @name = name
            @type_params = type_params
            @extension_name = extension_name
            @members = members
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(Extension) &&
              other.name == name &&
              other.type_params == type_params &&
              other.extension_name == extension_name &&
              other.members == members
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type_params.hash ^ extension_name.hash ^ members.hash
          end

          def to_json(*a)
            {
              declaration: :extension,
              name: name,
              type_params: type_params,
              extension_name: extension_name,
              members: members,
              annotations: annotations,
              location: location
            }.to_json(*a)
          end
        end

        class Interface
          attr_reader :name
          attr_reader :type_params
          attr_reader :members
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, type_params:, members:, annotations:, location:)
            @name = name
            @type_params = type_params
            @members = members
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(Interface) &&
              other.name == name &&
              other.type_params == type_params &&
              other.members == members
          end

          alias eql? ==

          def hash
            self.class.hash ^ type_params.hash ^ members.hash
          end

          def to_json(*a)
            {
              declaration: :interface,
              name: name,
              type_params: type_params,
              members: members,
              annotations: annotations,
              location: location
            }.to_json(*a)
          end
        end

        class Alias
          attr_reader :name
          attr_reader :type
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, type:, annotations:, location:)
            @name = name
            @type = type
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(Alias) &&
              other.name == name &&
              other.type == type
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type.hash
          end

          def to_json(*a)
            {
              declaration: :alias,
              name: name,
              type: type,
              annotations: annotations,
              location: location
            }.to_json(*a)
          end
        end

        class Constant
          attr_reader :name
          attr_reader :type
          attr_reader :location

          def initialize(name:, type:, location:)
            @name = name
            @type = type
            @location = location
          end

          def ==(other)
            other.is_a?(Constant) &&
              other.name == name &&
              other.type == type
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type.hash
          end

          def to_json(*a)
            {
              declaration: :constant,
              name: name,
              type: type,
              location: location
            }.to_json(*a)
          end
        end

        class Global
          attr_reader :name
          attr_reader :type
          attr_reader :location

          def initialize(name:, type:, location:)
            @name = name
            @type = type
            @location = location
          end

          def ==(other)
            other.is_a?(Global) &&
              other.name == name &&
              other.type == type
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type.hash
          end

          def to_json(*a)
            {
              declaration: :global,
              name: name,
              type: type,
              location: location
            }.to_json(*a)
          end
        end
      end
    end
  end
end

File: rust/ruby-rbs/src/ast/declarations.rs
use crate::ast::annotation::Annotation;
use crate::ast::comment::Comment;
use crate::ast::location::{
    AliasDeclarationLocation, ClassDeclarationLocation, ClassSuperLocation,
    ConstantDeclarationLocation, GlobalDeclarationLocation, InterfaceDeclarationLocation,
    ModuleDeclarationLocation, ModuleSelfLocation, TypeAliasDeclarationLocation,
};
use crate::ast::members::Member;
use crate::ast::type_param::TypeParam;
use crate::ast::types::Type;
use crate::ids::{SymbolId, TypeName};

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum Declaration {
    Class(ClassDeclaration),
    Module(ModuleDeclaration),
    Interface(InterfaceDeclaration),
    Constant(ConstantDeclaration),
    Global(GlobalDeclaration),
    TypeAlias(TypeAliasDeclaration),
    ClassAlias(ClassAliasDeclaration),
    ModuleAlias(ModuleAliasDeclaration),
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum ClassMember {
    Member(Member),
    Declaration(Declaration),
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum ModuleMember {
    Member(Member),
    Declaration(Declaration),
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ClassSuper {
    pub name: TypeName,
    pub args: Vec<Type>,
    pub location: Option<ClassSuperLocation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ClassDeclaration {
    pub name: TypeName,
    pub type_params: Vec<TypeParam>,
    pub members: Vec<ClassMember>,
    pub super_class: Option<ClassSuper>,
    pub annotations: Vec<Annotation>,
    pub location: Option<ClassDeclarationLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ModuleSelf {
    pub name: TypeName,
    pub args: Vec<Type>,
    pub location: Option<ModuleSelfLocation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ModuleDeclaration {
    pub name: TypeName,
    pub type_params: Vec<TypeParam>,
    pub members: Vec<ModuleMember>,
    pub location: Option<ModuleDeclarationLocation>,
    pub annotations: Vec<Annotation>,
    pub self_types: Vec<ModuleSelf>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct InterfaceDeclaration {
    pub name: TypeName,
    pub type_params: Vec<TypeParam>,
    pub members: Vec<Member>,
    pub annotations: Vec<Annotation>,
    pub location: Option<InterfaceDeclarationLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct TypeAliasDeclaration {
    pub name: TypeName,
    pub type_params: Vec<TypeParam>,
    pub ty: Type,
    pub annotations: Vec<Annotation>,
    pub location: Option<TypeAliasDeclarationLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ConstantDeclaration {
    pub name: TypeName,
    pub ty: Type,
    pub location: Option<ConstantDeclarationLocation>,
    pub comment: Option<Comment>,
    pub annotations: Vec<Annotation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct GlobalDeclaration {
    pub name: SymbolId,
    pub ty: Type,
    pub location: Option<GlobalDeclarationLocation>,
    pub comment: Option<Comment>,
    pub annotations: Vec<Annotation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ClassAliasDeclaration {
    pub new_name: TypeName,
    pub old_name: TypeName,
    pub location: Option<AliasDeclarationLocation>,
    pub comment: Option<Comment>,
    pub annotations: Vec<Annotation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ModuleAliasDeclaration {
    pub new_name: TypeName,
    pub old_name: TypeName,
    pub location: Option<AliasDeclarationLocation>,
    pub comment: Option<Comment>,
    pub annotations: Vec<Annotation>,
}
